app/models/mixins/purging_mixin.rb

```diff
@@ -46,6 +60,57 @@ def purge_by_date(older_than, window = nil, &block)
       total
     end
 
+    def purge_count_by_remaining(remaining)
+      purge_ids_for_remaining(remaining).size
+    end
+
+    # @param [Integer] remaining number of records per resource to keep (remains in table)
+    # @param [Integer] window number of records to delete in a batch
+    # @return [Integer] number of records deleted
+    def purge_by_remaining(remaining, window = nil, &block)
+      _log.info("Purging #{table_name.humanize} older than last #{remaining} results...")
+      total = purge_in_batches(purge_ids_for_remaining(remaining), window || purge_window_size, &block)
+      _log.info("Purging #{table_name.humanize} older than last #{remaining} results...Complete - Deleted #{total} records")
+      total
+    end
+
+    # Private:  The ids to purge if we want to keep a fixed number of records
+    # for each resource. Newer records (with a higher id/ lower rank) are kept.
+    #
+    # example:
+    #
+    # So given the following DriftStates:
+    #
+    # {id: 1, resource_type: 'foo', resource_id: 1}  # Rank = 2
+    # {id: 2, resource_type: 'foo', resource_id: 1}  # Rank = 1
+    # {id: 3, resource_type: 'foo', resource_id: 2}  # Rank = 2
+    # {id: 4, resource_type: 'foo', resource_id: 2}  # Rank = 1
+    # {id: 5, resource_type: 'bar', resource_id: 6}  # Rank = 1
+    # {id: 6, resource_type: 'baz', resource_id: 4}  # Rank = 1
+    #
+    # For a "remaining" value of `1`, id 1 and 3 would then be up for deletion.
+    #
+    # @param [Integer] remaining number of records per resource to keep (remains in table)
+    # @return [Scope] Records to be deleted
+    def purge_ids_for_remaining(remaining)
+      # NOTE: `as(table_name)` fixes a bug with pluck(:id)
+      from(purge_ids_ranked_by_age.as(table_name)).where("rank > ?", remaining)
+    end
+
+    # Private: Rank records by age for each resource
+    #
+    # Assigns a "RANK" value to each record, categorized by their foreign key(s).
+    #
+    # The higher the id, the lower the rank value, the higher the priority.
+    #
+    # The values to purge are the records with a rank greater than the number
+    # of records to keep.
+    #
+    # @return [Scope] Scope of the ids prioritized by age.
+    def purge_ids_ranked_by_age
+      select(%{"id", RANK() OVER(PARTITION BY #{purge_remaining_foreign_key} ORDER BY "id" DESC) as "rank"})
```

Comment:
A few things here:

- This might be easier to read as multi-line HEREDOC, maybe like the following.  I'm not sure, it's really up to you; I'm just trying to think about others (including myself) reading this in the future.

    ```ruby
    select(<<-EOSQL)
      "id", RANK() OVER(
        PARTITION BY #{purge_remaining_foreign_key}
        ORDER BY "id" DESC
      ) AS "rank"
    EOSQL
    ```

- I really don't like the SQL embedding with raw SQL, and expecting the caller to give the SQL fragment.  It might be nicer to have the caller give a set of symbols and then you loop over them and enquote them here.

    ```ruby
    # caller
    def purge_remaining_foreign_key
      [:resource_type, :resource_id]
    end

    # here
    partition_key = Array.wrap(purge_remaining_foreign_key).collect { |k| quote(k) }.join(", ")
    select(%{... #{partition_key} ...})
    ```

- Very minor, but capitalize the `AS` to conform to capitalizing SQL keywords as done elsewhere